### PR TITLE
Fix regex_replace

### PR DESCRIPTION
**Pull request by knu**
As discussed in #886, there are a couple of flaws in Liquid for the regex_replace/regex_replace_first filters to work as most users would expect.

- Use of the closing curly brace in string literals causes SyntaxError: Fix by monkey-patching Liquid
- No escaped characters supported in string literals: Fix by specially handling the replacement parameter of regex_replace/regex_replace_first

### CODE REVIEW

File: app/concerns/liquid_interpolatable.rb [hunk 1]
```diff
@@ -221,6 +221,43 @@ def logger
...
+      "3" => BACKSLASH + "3",
+      "4" => BACKSLASH + "4",
+      "5" => BACKSLASH + "5",
+      "6" => BACKSLASH + "6",
+      "7" => BACKSLASH + "7",
+      "8" => BACKSLASH + "8",
+      "9" => BACKSLASH + "9",
+    }
+
+    def unescape_replacement(s)
```
Comment (cantino) [inline app/concerns/liquid_interpolatable.rb]: This probably warrants an explanatory comment.

Comment (HyShai) [inline app/concerns/liquid_interpolatable.rb]: if possible I would also like to understand this code. Some interesting stuff going on here (to a ruby noob).

Comment (cantino) [inline app/concerns/liquid_interpolatable.rb]: It's running a gsub across the input, replacing every matching section with the contents returned from the block.  `$1`, `$2`, etc. are the first and second match groups copied into thread-local variables.  `to_i(16)` is converting into hex and [pack](http://ruby-doc.org/core-2.2.2/Array.html#method-i-pack) appears to be converting to unicode or ASCII, depending on whether it's reading a `\udddd` or `\xdd` input.

Comment (HyShai) [inline app/concerns/liquid_interpolatable.rb]: Interesting - why convert to hex and `pack`?

**Comment by knu** [comment]
I'm going to add something like the following to Wiki after merging this.

> You can use escape sequences in both the regex and the replacement parameter of these filter functions, such as `\r`, `\n`, `\xXX` and `\u{XXXX}`.  In the replacement parameter, numbered capture groups `\1`..`\9` are supported, as well as named capture groups `\k<name>`, `\&`, `` \` ``, `\'`, etc.
> Beware that Liquid itself does not have the concept of backslash escaping, so you cannot escape quote characters like `"\""`.  Instead, use `\x22` when you need to put the double quote character in a double quoted string literal, such as `{{ text | regex_replace: "\x22(.*?)\x22", "\1" }}`.

**Comment by HyShai** [comment]
Ah now I see why you're using hex. Thanks so much!!

**Comment by knu** [comment]
Squashed.

**Comment by cantino** [comment]
Looks good to me @knu!

**Pull request merged**